test/cpp/end2end/grpc_authz_end2end_test.cc

```diff
@@ -415,6 +485,253 @@ TEST_F(GrpcAuthzEnd2EndTest,
   EXPECT_EQ(resp.message(), kMessage);
 }
 
+TEST_F(GrpcAuthzEnd2EndTest, StaticInitWithAuditLoggingOnDeny) {
+  std::string policy =
+      "{"
+      "  \"name\": \"authz\","
+      "  \"allow_rules\": ["
+      "    {"
+      "      \"name\": \"allow_foo\","
+      "      \"request\": {"
+      "        \"headers\": ["
+      "          {"
+      "            \"key\": \"key-foo\","
+      "            \"values\": [\"foo\"]"
+      "          }"
+      "        ]"
+      "      }"
+      "    }"
+      "  ],"
+      "  \"deny_rules\": ["
+      "    {"
+      "      \"name\": \"deny_bar\","
+      "      \"request\": {"
+      "        \"headers\": ["
+      "          {"
+      "            \"key\": \"key-bar\","
+      "            \"values\": [\"bar\"]"
+      "          }"
+      "        ]"
+      "      }"
+      "    }"
+      "  ],"
+      "  \"audit_logging_options\": {"
+      "    \"audit_condition\": \"ON_DENY\","
+      "    \"audit_loggers\": ["
+      "      {"
+      "        \"name\": \"test_logger\""
+      "      },"
+      "      {"
+      "        \"name\": \"stdout_logger\""
+      "      }"
+      "    ]"
+      "  }"
+      "}";
+  InitServer(CreateStaticAuthzPolicyProvider(policy));
+  auto channel = BuildChannel();
+  grpc::testing::EchoResponse resp;
+  grpc::Status status;
+
+  ClientContext context1;
+  context1.AddMetadata("key-foo", "foo");
+  status = SendRpc(channel, &context1, &resp);
+  EXPECT_TRUE(status.ok());
+  EXPECT_EQ(log_count_, 0);
+
+  ClientContext context2;
+  context2.AddMetadata("key-foo", "bar");
+  ::testing::internal::CaptureStdout();
+  status = SendRpc(channel, &context2, &resp);
+  CompareAuditLogEntry(
+      ::testing::internal::GetCapturedStdout(),
+      "{\"authorized\":false,\"matched_rule\":\"\",\"policy_name\":\"authz\","
+      "\"principal\":\"spiffe://foo.com/bar/baz\",\"rpc_method\":\"/"
+      "grpc.testing.EchoTestService/Echo\"}");
+  EXPECT_EQ(status.error_code(), grpc::StatusCode::PERMISSION_DENIED);
+  EXPECT_EQ(status.error_message(), "Unauthorized RPC request rejected.");
+  EXPECT_EQ(log_count_, 1);
+
+  ClientContext context3;
```

Comment:
Would it be problematic to split this into 3 different tests? That way you don't have to chain failures or spend time figuring out which part failed

Reply:
I don't see any benefit from splitting them. When it fails somewhere, the test log will clearly show which one it is including the line number. I haven't had much trouble navigating that.